File: fashion_mnist_cnn.py
import os
import numpy as np
from tensorflow import keras
from tensorflow.keras.datasets import fashion_mnist

# Get the directory where this script is located
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

def build_cnn_model():
    # Create a sequential CNN model with 6 layers
    model = keras.Sequential([
        # First convolutional layer with 32 filters
        keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
        # First max pooling layer
        keras.layers.MaxPooling2D((2,2)),
        # Second convolutional layer with 64 filters
        keras.layers.Conv2D(64, (3,3), activation='relu'),
        # Second max pooling layer
        keras.layers.MaxPooling2D((2,2)),
        # Flatten layer to prepare for dense layers
        keras.layers.Flatten(),
        # Dense layer with 64 units
        keras.layers.Dense(64, activation='relu'),
        # Output layer with 10 units (one per class)
        keras.layers.Dense(10, activation='softmax')
    ])
    
    # Compile the model with appropriate settings
    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    return model

def main():
    # Load and preprocess Fashion MNIST data
    (x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
    
    # Normalize pixel values and add channel dimension
    x_train = x_train[..., None].astype('float32') / 255  # [None] adds channel dimension
    x_test = x_test[..., None].astype('float32') / 255

    # Build and train the CNN model
    model = build_cnn_model()
    print("Training model...")
    model.fit(x_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

    # Create full paths for saving assets
    model_path = os.path.join(SCRIPT_DIR, 'fashion_mnist_cnn.keras')
    test_images_path = os.path.join(SCRIPT_DIR, 'test_images.npy')
    test_labels_path = os.path.join(SCRIPT_DIR, 'test_labels.npy')

    # Save model and test data
    model.save(model_path)
    np.save(test_images_path, x_test)
    np.save(test_labels_path, y_test)
    print(f"Model and test data saved in: {SCRIPT_DIR}")

if __name__ == "__main__":
    main()

File: fashion_mnist_cnn.r
install.packages("this.path")
library(keras3)
library(this.path)

# Get directory of THIS script
script_dir <- this.dir()

# Set working directory to script location
setwd(script_dir)

build_model <- function() {
  keras_model_sequential(name = "fashion_mnist_cnn") %>%
    layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = "relu",
                  input_shape = c(28, 28, 1)) %>%
    layer_max_pooling_2d(pool_size = c(2, 2)) %>%
    layer_conv_2d(filters = 64, kernel_size = c(3, 3), activation = "relu") %>%
    layer_max_pooling_2d(pool_size = c(2, 2)) %>%
    layer_flatten() %>%
    layer_dense(units = 64, activation = "relu") %>%
    layer_dense(units = 10, activation = "softmax") %>%
    compile(
      optimizer = optimizer_adam(),
      loss = loss_sparse_categorical_crossentropy(),
      metrics = metric_sparse_categorical_accuracy()
    )
}

main <- function() {
  # Load data
  fashion_mnist <- dataset_fashion_mnist()
  x_train <- fashion_mnist$train$x
  y_train <- fashion_mnist$train$y
  x_test <- fashion_mnist$test$x
  y_test <- fashion_mnist$test$y

  # Preprocess
  x_train <- array_reshape(x_train, c(dim(x_train)[1], 28, 28, 1)) / 255
  x_test <- array_reshape(x_test, c(dim(x_test)[1], 28, 28, 1)) / 255

  # Build and train model
  model <- build_model()
  model %>% fit(
    x_train, y_train,
    epochs = 10,
    batch_size = 128,
    validation_split = 0.2
  )
  # Save to script directory
  model %>% save_model("fashion_mnist_cnn.keras")
  saveRDS(x_test, "test_images.rds")
  saveRDS(y_test, "test_labels.rds")
  message("Files saved in: ", getwd())
}

main()